### ISSUE TITLE
Automatically detect glyphs that need the overlap bits 

### ISSUE THREAD

**Issue by MariannaPaszkowska** [googlefonts/googlesans-flex#798]
related #737

**Comment by chrissimpkins** [comment]
@rsheeter how do you suggest that we approach this in Google Sans Flex?  Unset all bits and let internal pipeline handle it?  Does that approach have consequences for areas that our internal pipeline does not handle like Android system, GMS core, Figma corp distribution?

**Comment by chrissimpkins** [comment]
Some solution here will be a priority for v3.0, but it might not be the solution as described in the title.  

@MariannaPaszkowska @belluzj Will you please document the current approach that we use in GS Flex?  We are selectively setting bits for glyphs with known overlaps, correct?

**Comment by RickyDaMa** [comment]
The current approach is that there's a file maintained will all glyphs with overlap: `sources/glyphs-with-overlap.txt`

This file is used by `scripts/set-overlap-bits.py`, which is called as part of the postprocessing in the `Makefile` (i.e. after gftools has built the font): https://github.com/googlefonts/googlesans-flex/blob/d59e6d60b4f879b5e40544c088dbef0919727483/Makefile#L28

**Comment by rsheeter** [comment]
To ask the obvious, why don't we set it in sources?

**Comment by anthrotype** [comment]
fontmake should already support this glyph.lib key for DS+UFO projects (which I think is the case for GSF): https://unifiedfontobject.org/versions/ufo3/glyphs/glif/#publictruetypeoverlap

there's no equivalent Glyphs.app property yet AFAIK

**Comment by anthrotype** [comment]
this was added a couple years ago with ufo2ft [2.31.0](https://github.com/googlefonts/ufo2ft/releases/tag/v2.31.0)
https://github.com/googlefonts/ufo2ft/blob/d15ef8f6f235804733c6e2ea5f01f39aafb92f45/Lib/ufo2ft/instructionCompiler.py#L207-L212

**Comment by schriftgestalt** [comment]
We could add `public.truetype.overlap` to `glyph.userData` and make sure that glyphsLib picks it up.

**Comment by anthrotype** [comment]
I think that should already work, as the userData should be passed through as is. But there's no UI for setting that, you'd have to resort to a python script within the Glyphs.app Macro Panel.

**Comment by schriftgestalt** [comment]
wouldn’t you set that by a script anyway?

**Comment by anthrotype** [comment]
embedding in the source using the userData is doable via a script, but then the data is not visible at the UI level. At least the text file is visible and easily editable. In an ideal world Glyhps.app would have a checkbox for that somewhere.

**Comment by schriftgestalt** [comment]
I’ll think about this. There are many possible settings and I don’t like to add checkboxes for all of them. So this needs to be carefully considered.

**Comment by chrissimpkins** [comment]
I spoke with Rod about our overlap bit flag setting.  It is ok for us to continue to *selectively* set overlap bit flags on glyphs in the brand font projects.  It is not desirable to transition to global bit flag setting whether there are overlaps are not. 

He recommended an even-odd vs. non-zero fill rule SVG image prep + image diff approach to automate an understanding of whether there are overlaps in glyphs.  He developed the Py script in https://github.com/rsheeter/glyph_overlaps to produce the SVGs for such a comparison.  We'd probably need to do this at each master?

**Comment by rsheeter** [comment]
Nit: that python is exploratory, there might be reasons that approach doesn't work. Concept is that since the problem occurs when the render differs between fill rules directly check if that's true.

**Comment by anthrotype** [comment]
> Concept is that since the problem occurs when the render differs between fill rules directly check if that's true

Yeah, I think that's a good heuristic, thanks Rod!
By the way, 90% of that script of yours is parsing GLIF files and converting them to SVG. You can reduce that to a couple of lines by using fontTools.ufoLib and fontTools.pens.svgPathPen. It will take care of validation, decomposing composite glyphs, and all the edge cases.

```python
from fontTools.ufoLib import UFOReader
from fontTools.pens.svgPathPen import SVGPathPen

reader = UFOReader("path/to/font.ufo")
glyphs = reader.getGlyphSet()
for name in glyphs.keys():
    glyph = glyphs[name]
    pen = SVGPathPen(glyphs)
    glyph.draw(pen)
    path_d = pen.getCommands()
    print(path_d)
```

**Comment by rsheeter** [comment]
> You can reduce that to a couple of lines by using fontTools.ufoLib and fontTools.pens.svgPathPen

Good call, definitely want this or to rewrite in rust if we think this is actually useful.

**Comment by anthrotype** [comment]
maybe you could use Pillow's ImageChop.difference() to compare the two evenodd vs nonzero images

**Comment by anthrotype** [comment]
> rewrite in rust if we think this is actually useful

+1, and then we can call resvg directly instead of as external subprocess

**Comment by rsheeter** [comment]
Also we probably don't actually need to image diff. Brute force, just compute winding both ways for a lot of points (akin to rendering and comparing pixels). Probably there is a smarter way.

**Comment by anthrotype** [comment]
but then you end up implementing your own rasteriser.. the nice thing about using off-the-shelf solution is that we don't have to do that. we can trust resvg knows how correctly rasterise an svg with both fillrules, and we can just diff the result.

**Comment by rsheeter** [comment]
I don't need a full rasterizer, I just need the ability to compute winding for an arbitrary point. I think https://docs.rs/kurbo/latest/kurbo/struct.BezPath.html#method.winding might suffice?

**Comment by anthrotype** [comment]
https://github.com/fonttools/fonttools/blob/main/Lib/fontTools/pens/pointInsidePen.py supports both evenodd and nonzero fills

**Comment by rsheeter** [comment]
Consider me nerd-sniped I guess. I have to finish a couple chores and then I'm going to try it in Rust.

**Comment by chrissimpkins** [comment]
New info + exploration from Rod and Raph re: approaches to contour overlap detection

https://github.com/rsheeter/glyph_overlaps
